Sources/Identity/CustomerInfo.swift

```diff
@@ -164,263 +153,170 @@ import Foundation
-        case .firstSeenMissing, .firstSeenFormat:
-            let firstSeen = subscriberDictionary["first_seen"] as? String ?? "missing"
-            extraContext = "first_seen is: \(firstSeen)"
+        init(response: CustomerInfoResponse, schemaVersion: String?) {
+            self.response = response
+            self.schemaVersion = schemaVersion
         }
-        return CustomerInfoError.cantInstantiateJsonObject.addingUnderlyingError(subscriberDataError,
-                                                                                 extraContext: extraContext)
-    }
 
-    func activeKeys(dates: [String: Date?]) -> Set<String> {
-        return Set(dates.keys.filter {
-            guard let nonNullDate = dates[$0] as? Date else { return true }
-            return isAfterReferenceDate(date: nonNullDate)
-        })
     }
 
-    func isAfterReferenceDate(date: Date) -> Bool { date.timeIntervalSince(self.requestDate) > 0 }
+}
+
+/// `Codable` implementation that puts the content of`response` and `schemaVersion`
+/// at the same level instead of nested.
+extension CustomerInfo.Data: Codable {
+
+    private enum CodingKeys: String, CodingKey {
+
+        case response
+        case schemaVersion
 
-    func parseExpirationDates(transactionsByProductId: [String: [String: Any]]) -> [String: Date?] {
-        return parseDatesIn(transactionsByProductId: transactionsByProductId, dateLabel: "expires_date")
     }
 
-    func parsePurchaseDates(transactionsByProductId: [String: [String: Any]]) -> [String: Date?] {
-        return parseDatesIn(transactionsByProductId: transactionsByProductId, dateLabel: "purchase_date")
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+
+        try self.response.encode(to: encoder)
+        // Always use current schema version when encoding
+        try container.encode(CustomerInfo.currentSchemaVersion, forKey: .schemaVersion)
     }
 
-    func parseDatesIn(transactionsByProductId: [String: [String: Any]], dateLabel: String) -> [String: Date?] {
-        // mapValues will keep the key-value pair in the dictionary for nil values, as desired
-        return transactionsByProductId.mapValues { transaction in
-            if let dateString = transaction[dateLabel] as? String {
-                return dateFormatter.date(from: dateString)
-            }
-            return nil
-        }
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+
+        self.response = try CustomerInfoResponse(from: decoder)
+        self.schemaVersion = try container.decodeIfPresent(String.self, forKey: .schemaVersion)
     }
 
 }
 
-// Fixme: delete when `CustomerInfo` is `Decodable`
- extension CustomerInfo: HTTPResponseBody {
+private extension CustomerInfo {
+
+    func activeKeys(dates: [String: Date?]) -> Set<String> {
+        return Set(
+            dates
+                .lazy
```

Comment:
why is `lazy` needed here?

Reply:
it seems like gh posted my comments once per time I edited them before posting the review? I find a different but in gh for every PR over 700 lines.